processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java

```diff
@@ -1129,10 +1147,65 @@ public int add(String value)
     }
 
     @Override
+    public int compare(int lhsIdx, int rhsIdx)
+    {
+      if (cache != null && lhsIdx < capacity && rhsIdx < capacity) {
+        final boolean leftToRight = lhsIdx > rhsIdx;
+        final int[] cacheIndex = toIndex(lhsIdx, rhsIdx, leftToRight);
+        byte cached = getCache(cacheIndex);
+        if (cached == 0) {
+          final int compare = delegate.compare(lhsIdx, rhsIdx);
+          setCache(cacheIndex, cached = normalize(compare, leftToRight));
+        }
+        return leftToRight ? cached : -cached;
+      }
+      return delegate.compare(lhsIdx, rhsIdx);
+    }
+
+    @Override
     public SortedDimLookup sort()
     {
       return new NullValueConverterDimLookup(delegate.sort());
     }
+
+    private static final int[] MASKS = new int[]{
+        0b00000011, 0b00001100, 0b00110000, 0b11000000
+    };
+
+    private static final byte POSITIVE = (byte) 1;
+    private static final byte NEGATIVE = (byte) -1;
+
+    private static final byte ENCODED_POSITIVE = 0x01;
+    private static final byte ENCODED_NEGATIVE = 0x02;
+
+    private int[] toIndex(int lhsIdx, int rhsIdx, boolean leftToRight)
+    {
+      final int index = leftToRight ?
+                        (lhsIdx - rhsIdx - 1) + rhsIdx * (capacity - 1) - (rhsIdx * (rhsIdx - 1) / 2) :
+                        (rhsIdx - lhsIdx - 1) + lhsIdx * (capacity - 1) - (lhsIdx * (lhsIdx - 1) / 2);
+
+      return new int[]{index / 4, index % 4};
+    }
+
+    private byte getCache(int[] cacheIndex)
+    {
+      int encoded = (cache[cacheIndex[0]] & MASKS[cacheIndex[1]]) >>> (cacheIndex[1] << 1);
+      return encoded == ENCODED_POSITIVE ? POSITIVE : encoded == ENCODED_NEGATIVE ? NEGATIVE : 0;
+    }
+
+    // write once
+    private void setCache(int[] cacheIndex, byte value)
+    {
+      int encoded = value == POSITIVE ? ENCODED_POSITIVE : value == NEGATIVE ? ENCODED_NEGATIVE : 0;
+      if (encoded != 0) {
+        cache[cacheIndex[0]] |= (byte)(encoded << (cacheIndex[1] << 1));
+      }
+    }
+
+    private byte normalize(int compare, boolean leftToRight)
+    {
+      return leftToRight ^ compare > 0 ? NEGATIVE : POSITIVE;
+    }
```

Comment:
I've ran druid tests with hard coded cache size and passed all. agree on it's demanding serious test cases. I'll try